src/hotspot/share/gc/g1/heapRegionManager.cpp

```diff
@@ -174,13 +166,26 @@ HeapRegion* HeapRegionManager::new_heap_region(uint hrm_index) {
   return g1h->new_heap_region(hrm_index, mr);
 }
 
+void HeapRegionManager::expand(uint start, uint num_regions, WorkGang* pretouch_gang) {
+  commit_regions(start, num_regions, pretouch_gang);
+  for (uint i = start; i < start + num_regions; i++) {
+    HeapRegion* hr = _regions.get_by_index(i);
+    if (hr == NULL) {
+      hr = new_heap_region(i);
+      OrderAccess::storestore();
+      _regions.set_by_index(i, hr);
+      _allocated_heapregions_length = MAX2(_allocated_heapregions_length, i + 1);
+    }
+    G1CollectedHeap::heap()->hr_printer()->commit(hr);
+  }
+  activate_regions(start, num_regions);
```

Comment:
In this place there will be two messages from the HRPrinter for every region:
1) a COMMIT message and
2) an ACTIVATE message

This is a bit confusing as in my understanding (that's why I asked for a region state diagram in the `G1CommittedRegionMap` which may as well be put in `HeapRegionManager`) the (typical) flow of states are Uncommitted->Committed/Active->Committed/Inactive->Uncommitted.
As mentioned, I'm not sure if it is a good idea to send two separate messages here; better rename the "Active" message to "Commit-Active" (and "Inactive" to "Commit-Inactive") instead imho, even if it's quite long (and drop `HRPrinter::commit()` completely)

Reply:
It's true that it will generate two messages when committing a previously uncommitted region, but I still think it is valuable to separate them since we can also have the state change `Active->Inactive->Active`. In this case the transition from Inactive to Active will not include a commit, but rather making inactive regions active again. Just seeing a "Commit-Active" message in this case would not be as clear as seeing "Active" that is not immediately preceded by "Commit".

Reply:
Okay, I looked at the `G1HRPrinter` again, and indeed it prints the action, not the region state, so this is a fair point.